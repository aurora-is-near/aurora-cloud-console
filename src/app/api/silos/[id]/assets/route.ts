import { snakeCase } from "change-case"
import { abort } from "@/utils/abort"
import { toError } from "@/utils/errors"
import { createApiEndpoint } from "@/utils/api"
import { createStorageClient } from "@/supabase/create-storage-client"
import { updateSilo } from "@/actions/silos/update-silo"
import { SiloAsset } from "@/types/assets"

export const POST = createApiEndpoint("uploadSiloAsset", async (_req, ctx) => {
  const storage = createStorageClient()
  const file = ctx.body.get("file") as File
  const type = ctx.body.get("type") as SiloAsset

  if (!file) {
    abort(400, "No file provided")
  }

  const filename = snakeCase(file.name)
  const path = `/${ctx.params.id}/${type}sss/${filename}`

  const uploadRes = await storage.from("silo_assets").upload(path, file, {
    contentType: file.type,
    upsert: true,
  })

  if (uploadRes.error) {
    abort(500, toError(uploadRes.error).message)
  }

  const {
    data: { publicUrl: url },
  } = storage.from("silo_assets").getPublicUrl(path)

  await updateSilo(Number(ctx.params.id), {
    [type]: url,
  })

  return { url }
})
