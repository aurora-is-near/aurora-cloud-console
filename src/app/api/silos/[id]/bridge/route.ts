import { createApiEndpoint } from "@/utils/api"
import { abort } from "../../../../../utils/abort"
import { getSilo } from "@/actions/silos/get-silo"
import { getSiloBridge } from "@/actions/silo-bridge/get-silo-bridge"
import { createSiloBridge } from "@/actions/silo-bridge/create-silo-bridge"
import { Bridge, BridgeNetworkType } from "@/types/types"
import { BridgeSchema } from "@/types/api-schemas"
import { updateSiloBridge } from "@/actions/silo-bridge/update-silo-bridge"
import { isValidNetwork } from "@/utils/bridge"

const getValidNetworks = (networks: string[]): BridgeNetworkType[] => {
  const validNetworks = networks.filter(isValidNetwork)
  const invalidNetworks = networks.filter((network) => !isValidNetwork(network))

  if (invalidNetworks.length) {
    abort(400, `Invalid network(s): ${invalidNetworks.join(", ")}`)
  }

  return validNetworks
}

const getBridgeSchema = (bridge?: Bridge): BridgeSchema => {
  if (!bridge) {
    return {
      enabled: false,
      createdAt: null,
      updatedAt: null,
      fromNetworks: null,
      toNetworks: null,
      tokens: [],
    }
  }

  return {
    enabled: true,
    createdAt: bridge.created_at,
    updatedAt: bridge.updated_at,
    fromNetworks: bridge.from_networks,
    toNetworks: bridge.to_networks,
    tokens: bridge.tokens,
  }
}

export const GET = createApiEndpoint("getSiloBridge", async (_req, ctx) => {
  const bridge = await getSiloBridge(Number(ctx.params.id))

  if (!bridge) {
    return getBridgeSchema()
  }

  return getBridgeSchema(bridge)
})

export const POST = createApiEndpoint("createSiloBridge", async (_req, ctx) => {
  const silo = await getSilo(Number(ctx.params.id))

  if (!silo) {
    abort(404)
  }

  const existingBridge = await getSiloBridge(silo.id)

  if (existingBridge) {
    abort(400, "Bridge already exists")
  }

  const bridge = await createSiloBridge({ silo_id: silo.id })

  return getBridgeSchema(bridge)
})

export const PUT = createApiEndpoint("updateSiloBridge", async (_req, ctx) => {
  const siloId = Number(ctx.params.id)
  const bridge = await getSiloBridge(siloId)

  if (!bridge) {
    abort(404)
  }

  const { fromNetworks, toNetworks } = ctx.body

  const updatedBridge = await updateSiloBridge(siloId, {
    fromNetworks: fromNetworks ? getValidNetworks(fromNetworks) : undefined,
    toNetworks: toNetworks ? getValidNetworks(toNetworks) : undefined,
    tokens: ctx.body.tokens,
  })

  return getBridgeSchema(updatedBridge)
})
