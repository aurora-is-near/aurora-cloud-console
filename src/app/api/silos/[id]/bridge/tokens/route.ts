import { createApiEndpoint } from "@/utils/api"
import { abort } from "../../../../../../utils/abort"
import { getSilo } from "@/actions/silos/get-silo"
import { getSiloToken } from "@/actions/silo-tokens/get-silo-token"
import { updateToken } from "@/actions/tokens/update-token"
import { ApiResponseBody } from "@/types/api"
import { getSiloTokenByAddress } from "@/actions/silo-tokens/get-silo-token-by-address"
import { createToken } from "@/actions/tokens/create-token"

const bridgeToken = async (tokenId: number) => {
  const updatedToken = await updateToken(tokenId, {
    bridge_deployment_status: "PENDING",
  })

  return { status: updatedToken.bridge_deployment_status }
}

const bridgeExistingToken = async (
  siloId: number,
  tokenId: number,
): Promise<ApiResponseBody<"bridgeSiloToken">> => {
  const token = await getSiloToken(siloId, tokenId)

  if (!token) {
    abort(404)
  }

  if (token.bridge_deployment_status === "DEPLOYED") {
    abort(400, "Token is already deployed")
  }

  return bridgeToken(token.id)
}

const bridgeCustomToken = async (
  siloId: number,
  symbol: string,
  address: string,
): Promise<ApiResponseBody<"bridgeSiloToken">> => {
  const token = await getSiloTokenByAddress(siloId, address)

  if (token?.bridge_deployment_status === "DEPLOYED") {
    abort(400, "Token is already deployed")
  }

  if (token) {
    return bridgeToken(token.id)
  }

  const newToken = await createToken({
    symbol,
    address,
    deployment_status: "PENDING",
    bridge_deployment_status: "PENDING",
    silo_id: siloId,
    type: null,
    name: null,
    decimals: null,
    bridge_addresses: [],
    fast_bridge: false,
    bridge_origin: null,
    icon_url: null,
  })

  return { status: newToken.bridge_deployment_status }
}

export const POST = createApiEndpoint("bridgeSiloToken", async (_req, ctx) => {
  const silo = await getSilo(Number(ctx.params.id))
  const badRequestMessage = "Must provide either tokenId or symbol/address"

  if (!silo) {
    abort(404)
  }

  if (ctx.body.tokenId && (ctx.body.symbol || ctx.body.address)) {
    abort(400, badRequestMessage)
  }

  if (ctx.body.tokenId) {
    return bridgeExistingToken(silo.id, ctx.body.tokenId)
  }

  if (!ctx.body.symbol || !ctx.body.address) {
    abort(400, badRequestMessage)
  }

  return bridgeCustomToken(silo.id, ctx.body.symbol, ctx.body.address)
})
