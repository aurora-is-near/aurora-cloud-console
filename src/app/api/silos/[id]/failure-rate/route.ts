import { getSilo } from "@/actions/silos/get-silo"
import { abort } from "@/utils/abort"
import { createApiEndpoint } from "@/utils/api"
import { getFailureRateChart } from "@/utils/grafana/charts"

export const GET = createApiEndpoint(
  "getSiloFailureRate",
  async (_req, ctx) => {
    const silo = await getSilo(Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    const item = await getFailureRateChart({
      network: silo.grafana_network_key,
    })

    return { items: [item] }
  },
)
