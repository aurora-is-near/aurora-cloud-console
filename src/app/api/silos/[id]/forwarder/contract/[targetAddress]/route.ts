import { getAddress } from "ethers"
import { createApiEndpoint } from "@/utils/api"
import { abort } from "@/utils/abort"
import { forwarderApiClient } from "@/utils/forwarder-api/client"
import { getTeamSilo } from "@/actions/team-silos/get-team-silo"

export const GET = createApiEndpoint(
  "getForwarderAddress",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    let targetAddress

    try {
      targetAddress = getAddress(ctx.params.targetAddress)
    } catch {
      abort(400, "Invalid address")
    }

    const result = await forwarderApiClient.getContract({
      targetAddress,
      targetNetwork: silo.engine_account,
    })

    if (!result.result.address) {
      abort(404)
    }

    return {
      address: result.result.address,
    }
  },
)
