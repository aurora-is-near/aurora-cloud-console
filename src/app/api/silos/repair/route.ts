import { NextRequest } from "next/server"
import PQueue from "p-queue"
import { createPrivateApiEndpoint } from "@/utils/api"
import { deployDefaultTokens } from "@/actions/deployment/deploy-default-tokens"
import { getSilosToInspect } from "@/actions/silos/get-silos-to-inspect"
import {
  BaseTokenSymbol,
  Silo,
  SiloConfigTransactionStatus,
} from "@/types/types"
import { updateSilo } from "@/actions/silos/update-silo"
import { setBaseToken } from "@/actions/deployment/set-base-token"
import { getSiloBridgedTokenRequests } from "@/actions/silo-bridged-tokens/get-silo-bridged-token-requests"
import { getSiloBridgedTokens } from "@/actions/silo-bridged-tokens/get-silo-bridged-tokens"
import { resolveBridgedTokenRequest } from "@/actions/silo-bridged-tokens/resolve-bridged-token-request"
import { getBridgedTokens } from "@/actions/bridged-tokens/get-bridged-tokens"
import { createSiloBridgedToken } from "@/actions/silo-bridged-tokens/create-silo-bridged-token"
import { isBridgedTokenDeployed } from "@/utils/is-bridged-token-deployed"
import { updateSiloBridgedToken } from "@/actions/silo-bridged-tokens/update-silo-bridged-token"
import { abort } from "@/utils/abort"
import { AUTOMATED_BASE_TOKENS } from "@/constants/base-token"

type PreviouslyInspectedSilo = Omit<Silo, "inspected_at"> & {
  inspected_at: string
}

const queue = new PQueue({ concurrency: 3 })

const isPreviouslyInspectedSilo = (
  silo: Silo | PreviouslyInspectedSilo,
): silo is PreviouslyInspectedSilo => !!silo.inspected_at

/**
 * Check if any pending bridged tokens can be resolved.
 */
const resolvePendingBridgedTokens = async (silo: Silo) => {
  const [requests, siloBridgedTokens, bridgedTokens] = await Promise.all([
    getSiloBridgedTokenRequests(silo.id),
    getSiloBridgedTokens(silo.id),
    getBridgedTokens(),
  ])

  // Check if any requests for custom tokens can be resolved. If we have
  // subsequently configured a token with the same symbol as a request we can
  // resolve that request and assign the bridged token to the silo.
  await Promise.all(
    requests.map(async (request) => {
      if (siloBridgedTokens.some((token) => token.symbol === request.symbol)) {
        return
      }

      const bridgedToken = bridgedTokens.find(
        (token) => token.symbol === request.symbol,
      )

      if (!bridgedToken) {
        return
      }

      await createSiloBridgedToken(silo.id, bridgedToken.id, {
        isDeploymentPending: !(await isBridgedTokenDeployed(
          silo,
          bridgedToken,
        )),
      })

      await resolveBridgedTokenRequest(request.id)
    }),
  )

  // Check if any bridge tokens previously marked as pending have been deployed,
  // mark them as deployed if so.
  await Promise.all(
    siloBridgedTokens
      .filter((bridgedToken) => bridgedToken.is_deployment_pending)
      .map(async (bridgedToken) => {
        if (await isBridgedTokenDeployed(silo, bridgedToken)) {
          await updateSiloBridgedToken(silo.id, bridgedToken.id, {
            isDeploymentPending: false,
          })
        }
      }),
  )
}

const isAutomatableBaseToken = (baseToken?: string) =>
  AUTOMATED_BASE_TOKENS.includes(baseToken as BaseTokenSymbol)

/**
 * Perform various essential transactions to initialise a silo.
 */
const initialiseSilo = async (silo: PreviouslyInspectedSilo) => {
  const isWithin24Hours =
    Date.now() - new Date(silo.inspected_at).getTime() < 24 * 60 * 60 * 1000

  const transactionResults: SiloConfigTransactionStatus[] = await Promise.all([
    deployDefaultTokens(silo, { skipIfFailed: isWithin24Hours }),
    isAutomatableBaseToken(silo.base_token_symbol)
      ? setBaseToken(silo, { skipIfFailed: isWithin24Hours })
      : "PENDING",
  ])

  const siloUpdateProperties: Partial<Silo> = {
    inspected_at: new Date().toISOString(),
  }

  // If the silo has never been marked as active, mark it as such.
  if (
    !silo.is_active &&
    transactionResults.every((result) => result === "SUCCESSFUL")
  ) {
    siloUpdateProperties.is_active = true
  }

  await updateSilo(silo.id, siloUpdateProperties)
}

/**
 * Perform various checks and, if necessary, transactions to repair a silo.
 *
 * If the silo has never been inspected mark it as such and defer to the next
 * inspection. This is to help ensure we have a chance for the user to go
 * through the onboarding process without this check interrupting and taking
 * over before they have a chance to do so.
 */
const repairSilo = async (silo: Silo) => {
  if (!isPreviouslyInspectedSilo(silo)) {
    await updateSilo(silo.id, { inspected_at: new Date().toISOString() })

    return
  }

  await initialiseSilo(silo)
  await resolvePendingBridgedTokens(silo)
}

// This endpoint is intended to be called by a Vercel cron job, which only works
// with GET requests.
// https://vercel.com/docs/cron-jobs
export const GET = createPrivateApiEndpoint(async (req: NextRequest) => {
  if (req.headers.get("user-agent") !== "vercel-cron/1.0") {
    abort(403, "Forbidden")
  }

  const silos = await getSilosToInspect()

  await Promise.all(
    silos.map(async (silo) => {
      await queue.add(() => repairSilo(silo))
    }),
  )

  await queue.onIdle()

  return {
    status: 200,
    body: { message: "ok" },
  }
})
