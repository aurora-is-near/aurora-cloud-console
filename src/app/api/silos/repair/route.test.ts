/**
 * @jest-environment node
 */
import { NextRequest } from "next/server"
import { ethers } from "ethers"
import { contractChangerApiClient } from "@/utils/contract-changer-api/contract-changer-api-client"
import { logger } from "@/logger"
import { GET } from "./route"
import {
  createInsertOrUpdate,
  createSelect,
  mockSupabaseClient,
} from "../../../../../test-utils/mock-supabase-client"
import {
  createMockSilo,
  createMockSilos,
} from "../../../../../test-utils/factories/silo-factory"
import { createMockBridgedToken } from "../../../../../test-utils/factories/bridged-token-factory"
import { createMockBridgedTokenRequest } from "../../../../../test-utils/factories/silo-bridged-token-request-factory"

const currentTime = new Date("2025-03-17T09:00:00Z")

jest.useFakeTimers()
jest.setSystemTime(currentTime)

jest.mock("ethers")
jest.mock("@/utils/contract-changer-api/contract-changer-api-client", () => ({
  contractChangerApiClient: {
    mirrorErc20Token: jest.fn(({ token }) => ({
      tx_hash: `mock_tx_hash_${token}`,
    })),
    setBaseToken: jest.fn(() => ({ tx_hash: "mock_set_base_token_tx_hash" })),
  },
}))

jest.mock("near-api-js", () => ({
  ...jest.requireActual("near-api-js"),
  Account: jest.fn(() => ({
    getAccountBalance: jest.fn(() => ({
      total: "33137693971864085399999999",
      available: "18370803971864085399999999",
      staked: "0",
    })),
    viewFunction: jest.fn(() => ({
      total: "12345678901234567890",
      available: "42",
    })),
  })),
  connect: jest.fn(() => ({
    connection: {
      provider: {
        txStatus: jest.fn(() => ({ status: { SuccessValue: "" } })),
      },
    },
  })),
}))

describe("Silos repair route", () => {
  beforeEach(() => {
    mockSupabaseClient.from("silos").select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("silos")
      .update.mockReturnValue(createInsertOrUpdate({}))
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("silo_config_transactions")
      .update.mockReturnValue(createInsertOrUpdate([]))
    mockSupabaseClient
      .from("bridged_token_requests")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("bridged_tokens")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("silo_bridged_tokens")
      .select.mockReturnValue(createSelect([]))
    ;(ethers.Contract as jest.Mock).mockImplementation(() => ({
      symbol: () => {
        throw new Error("Not implemented")
      },
    }))
  })

  it("repairs all silos", async () => {
    const oneHourAgo = new Date(currentTime.getTime() - 60 * 60 * 1000)

    const mockSilos = createMockSilos(2, {
      is_active: false,
      inspected_at: oneHourAgo.toISOString(),
    })

    mockSilos[0].base_token_symbol = "AURORA"

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect(mockSilos))
    ;(ethers.Contract as jest.Mock).mockImplementation(
      (tokenContractAddress) => ({
        symbol: () => {
          if (
            tokenContractAddress ===
            "0xC42C30aC6Cc15faC9bD938618BcaA1a1FaE8501d"
          ) {
            throw new Error("Not implemented")
          }
        },
      }),
    )

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(ethers.Contract).toHaveBeenCalledTimes(8)
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledTimes(3)

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledWith({
      target: null,
      operation: "DEPLOY_NEAR",
      silo_id: mockSilos[0].id,
      status: "PENDING",
      transaction_hash: "mock_tx_hash_Near",
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledWith({
      target: null,
      operation: "DEPLOY_NEAR",
      silo_id: mockSilos[1].id,
      status: "PENDING",
      transaction_hash: "mock_tx_hash_Near",
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledWith({
      target: "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near",
      operation: "SET_BASE_TOKEN",
      silo_id: mockSilos[0].id,
      status: "PENDING",
      transaction_hash: "mock_set_base_token_tx_hash",
    })

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledTimes(2)
    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenNthCalledWith(
      1,
      {
        siloEngineAccountId: mockSilos[0].engine_account,
        token: "Near",
      },
    )

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenNthCalledWith(
      2,
      {
        siloEngineAccountId: mockSilos[1].engine_account,
        token: "Near",
      },
    )

    expect(contractChangerApiClient.setBaseToken).toHaveBeenCalledTimes(1)
    expect(contractChangerApiClient.setBaseToken).toHaveBeenCalledWith({
      baseTokenAccountId:
        "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near",
      siloEngineAccountId: mockSilos[0].engine_account,
    })

    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledTimes(2)
    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledWith({
      inspected_at: currentTime.toISOString(),
    })
  })

  it("defers repair of a silo that has never been inspected", async () => {
    const mockSilo = createMockSilo({
      is_active: false,
      base_token_symbol: "AURORA",
      inspected_at: null,
    })

    const silosUpdateQueries = createInsertOrUpdate(mockSilo)

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect([mockSilo]))
    mockSupabaseClient.from("silos").update.mockReturnValue(silosUpdateQueries)

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(contractChangerApiClient.setBaseToken).not.toHaveBeenCalled()
    expect(ethers.Contract).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()

    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledTimes(1)
    expect(silosUpdateQueries.eq).toHaveBeenCalledWith("id", mockSilo.id)
    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledWith({
      inspected_at: currentTime.toISOString(),
    })
  })

  it("does not perform update transactions if the last transaction failed and the inspection was within 24 hours", async () => {
    const oneDayAgo = new Date(currentTime.getTime() - 24 * 60 * 60 * 1000)

    const mockSilo = createMockSilo({
      is_active: false,
      base_token_symbol: "AURORA",
      inspected_at: new Date(oneDayAgo.getTime() + 1).toISOString(),
    })

    const silosUpdateQueries = createInsertOrUpdate(mockSilo)

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect([mockSilo]))
    mockSupabaseClient.from("silos").update.mockReturnValue(silosUpdateQueries)
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([{ status: "FAILED" }]))

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(contractChangerApiClient.setBaseToken).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()

    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledTimes(1)
    expect(silosUpdateQueries.eq).toHaveBeenCalledWith("id", mockSilo.id)
    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledWith({
      inspected_at: currentTime.toISOString(),
    })
  })

  it("marks a silo as active if all transactions were completed", async () => {
    const oneHourAgo = new Date(currentTime.getTime() - 60 * 60 * 1000)

    const mockSilo = createMockSilo({
      is_active: false,
      base_token_symbol: "AURORA",
      inspected_at: oneHourAgo.toISOString(),
    })

    const silosUpdateQueries = createInsertOrUpdate(mockSilo)

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect([mockSilo]))
    mockSupabaseClient.from("silos").update.mockReturnValue(silosUpdateQueries)
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([{ status: "SUCCESSFUL" }]))

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledTimes(1)
    expect(silosUpdateQueries.eq).toHaveBeenCalledWith("id", mockSilo.id)
    expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledWith({
      is_active: true,
      inspected_at: currentTime.toISOString(),
    })
  })

  it("resolves pending bridged token requests", async () => {
    const mockSilo = createMockSilo({
      is_active: true,
      base_token_symbol: "AURORA",
      inspected_at: new Date(
        currentTime.getTime() - 60 * 60 * 1000,
      ).toISOString(),
    })

    const mockBridgedTokenReqeusts = [
      createMockBridgedTokenRequest({ symbol: "TESTa" }),
      createMockBridgedTokenRequest({ symbol: "TESTb" }),
      createMockBridgedTokenRequest({ symbol: "TESTc" }),
    ]

    const mockBridgedTokens = [
      {
        ...createMockBridgedToken({ symbol: "TESTa", silo_address: "0x123" }),
        silo_bridged_tokens: [],
      },
      {
        ...createMockBridgedToken({ symbol: "TESTb", silo_address: "0x456" }),
        silo_bridged_tokens: [],
      },
    ]

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect([mockSilo]))
    mockSupabaseClient
      .from("bridged_token_requests")
      .select.mockReturnValue(createSelect(mockBridgedTokenReqeusts))
    mockSupabaseClient
      .from("bridged_tokens")
      .select.mockReturnValue(createSelect(mockBridgedTokens))
    ;(ethers.Contract as jest.Mock).mockImplementation(
      (tokenContractAddress) => ({
        symbol: () => {
          if (tokenContractAddress !== mockBridgedTokens[0].silo_address) {
            throw new Error("Not implemented")
          }
        },
      }),
    )

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(
      mockSupabaseClient.from("silo_bridged_tokens").upsert,
    ).toHaveBeenCalledTimes(2)
    expect(
      mockSupabaseClient.from("silo_bridged_tokens").upsert,
    ).toHaveBeenCalledWith([
      {
        bridged_token_id: mockBridgedTokenReqeusts[0].id,
        is_deployment_pending: true,
        silo_id: mockSilo.id,
      },
    ])
    expect(
      mockSupabaseClient.from("silo_bridged_tokens").upsert,
    ).toHaveBeenCalledWith([
      {
        bridged_token_id: mockBridgedTokenReqeusts[1].id,
        is_deployment_pending: false,
        silo_id: mockSilo.id,
      },
    ])
  })

  it("resolves pending bridged tokens", async () => {
    const mockSilo = createMockSilo({
      is_active: true,
      base_token_symbol: "AURORA",
      inspected_at: new Date(
        currentTime.getTime() - 60 * 60 * 1000,
      ).toISOString(),
    })

    const mockBridgedTokens = [
      {
        ...createMockBridgedToken({ symbol: "TESTa", silo_address: "0x123" }),
        silo_bridged_tokens: [
          { silo_id: mockSilo.id, is_deployment_pending: true },
        ],
      },
      {
        ...createMockBridgedToken({ symbol: "TESTb", silo_address: "0x456" }),
        silo_bridged_tokens: [
          { silo_id: mockSilo.id, is_deployment_pending: true },
        ],
      },
    ]

    const silosBridgedTokensUpdateQueries = createInsertOrUpdate({})

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect([mockSilo]))
    mockSupabaseClient
      .from("bridged_token_requests")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("bridged_tokens")
      .select.mockReturnValue(createSelect(mockBridgedTokens))
    mockSupabaseClient
      .from("silo_bridged_tokens")
      .update.mockReturnValue(silosBridgedTokensUpdateQueries)
    ;(ethers.Contract as jest.Mock).mockImplementation(
      (tokenContractAddress) => ({
        symbol: () => {
          if (tokenContractAddress !== mockBridgedTokens[0].silo_address) {
            throw new Error("Not implemented")
          }
        },
      }),
    )

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(
      mockSupabaseClient.from("silo_bridged_tokens").update,
    ).toHaveBeenCalledTimes(2)

    expect(silosBridgedTokensUpdateQueries.eq).toHaveBeenCalledWith(
      "silo_id",
      mockSilo.id,
    )

    expect(silosBridgedTokensUpdateQueries.eq).toHaveBeenCalledWith(
      "bridged_token_id",
      mockBridgedTokens[0].id,
    )

    expect(silosBridgedTokensUpdateQueries.eq).toHaveBeenCalledWith(
      "bridged_token_id",
      mockBridgedTokens[1].id,
    )

    expect(
      mockSupabaseClient.from("silo_bridged_tokens").update,
    ).toHaveBeenCalledWith({
      is_deployment_pending: false,
    })
  })

  it("returns a 403 if no valid user agent given", async () => {
    logger.error = jest.fn()

    const req = new NextRequest("https://example.com", {
      method: "GET",
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(403)
    expect(await res.json()).toEqual({
      statusCode: 403,
      message: "Forbidden",
      type: "/probs/forbidden",
    })

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(contractChangerApiClient.setBaseToken).not.toHaveBeenCalled()
    expect(mockSupabaseClient.from("silos").update).not.toHaveBeenCalled()
    expect(ethers.Contract).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()

    expect(logger.error).toHaveBeenCalledTimes(1)
    expect(logger.error).toHaveBeenCalledWith(new Error("Forbidden"))
  })

  it.each(["CUSTOM", "ART", "BTC"])(
    "does not perform a transaction or set the silo to active for base token %p",
    async (symbol) => {
      const oneHourAgo = new Date(currentTime.getTime() - 60 * 60 * 1000)

      const mockSilo = createMockSilo({
        base_token_symbol: symbol,
        is_active: false,
        inspected_at: oneHourAgo.toISOString(),
      })

      mockSupabaseClient
        .from("silos")
        .select.mockReturnValue(createSelect([mockSilo]))
      ;(ethers.Contract as jest.Mock).mockImplementation(() => ({
        symbol: () => {},
      }))

      const req = new NextRequest("https://example.com", {
        method: "GET",
        headers: {
          "user-agent": "vercel-cron/1.0",
        },
      })

      const res = await GET(req, { params: {} })

      expect(res.status).toBe(200)
      expect(await res.json()).toEqual({
        status: 200,
        body: {
          message: "ok",
        },
      })

      expect(ethers.Contract).toHaveBeenCalledTimes(5)
      expect(
        mockSupabaseClient.from("silo_config_transactions").insert,
      ).not.toHaveBeenCalled()

      expect(contractChangerApiClient.setBaseToken).not.toHaveBeenCalled()

      expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledTimes(1)
      expect(mockSupabaseClient.from("silos").update).toHaveBeenCalledWith({
        inspected_at: currentTime.toISOString(),
      })
    },
  )

  it("initialises the silo's set of bridged tokens", async () => {
    const oneHourAgo = new Date(currentTime.getTime() - 60 * 60 * 1000)

    const mockSilo = createMockSilo({
      base_token_symbol: "TURBO",
      is_active: false,
      inspected_at: oneHourAgo.toISOString(),
    })

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect([mockSilo]))
    ;(ethers.Contract as jest.Mock).mockImplementation(() => ({
      symbol: () => {},
    }))

    mockSupabaseClient.from("bridged_tokens").select.mockImplementation(() =>
      createSelect(
        ["AURORA", "ETH", "NEAR", "USDC", "USDt", "TURBO"].map(
          (symbol, index) => ({
            ...createMockBridgedToken({ id: 42 + index, symbol }),
            silo_bridged_tokens: [],
          }),
        ),
      ),
    )

    const req = new NextRequest("https://example.com", {
      method: "GET",
      headers: {
        "user-agent": "vercel-cron/1.0",
      },
    })

    const res = await GET(req, { params: {} })

    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({
      status: 200,
      body: {
        message: "ok",
      },
    })

    expect(
      mockSupabaseClient.from("silo_bridged_tokens").upsert,
    ).toHaveBeenCalledTimes(6)

    expect(
      mockSupabaseClient.from("silo_bridged_tokens").upsert.mock.calls,
    ).toMatchSnapshot()
  })
})
