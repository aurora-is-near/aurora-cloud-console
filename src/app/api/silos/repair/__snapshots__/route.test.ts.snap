// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Silos repair route initialises the silo's set of bridged tokens 1`] = `
[
  [
    [
      {
        "bridged_token_id": 47,
        "is_deployment_pending": true,
        "silo_id": 1,
      },
    ],
  ],
  [
    [
      {
        "bridged_token_id": 44,
        "is_deployment_pending": true,
        "silo_id": 1,
      },
    ],
  ],
  [
    [
      {
        "bridged_token_id": 42,
        "is_deployment_pending": true,
        "silo_id": 1,
      },
    ],
  ],
  [
    [
      {
        "bridged_token_id": 46,
        "is_deployment_pending": true,
        "silo_id": 1,
      },
    ],
  ],
  [
    [
      {
        "bridged_token_id": 45,
        "is_deployment_pending": true,
        "silo_id": 1,
      },
    ],
  ],
]
`;
